src/Evolution/Systems/ScalarWave/BoundaryConditions/ConstraintPreservingSphericalRadiation.cpp

```diff
@@ -0,0 +1,178 @@
+    const gsl::not_null<Scalar<DataVector>*> dt_pi_correction,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        dt_phi_correction,
+    const gsl::not_null<Scalar<DataVector>*> dt_psi_correction,
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+        face_mesh_velocity,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_covector,
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi,
+    const Scalar<DataVector>& psi,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& coords,
+    const Scalar<DataVector>& gamma2, const Scalar<DataVector>& dt_pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& dt_phi,
+    const Scalar<DataVector>& dt_psi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& d_pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& d_psi,
+    const tnsr::ij<DataVector, Dim, Frame::Inertial>& d_phi) const noexcept {
+  {
+    DataVector& inv_radius = get(*dt_psi_correction);
+    inv_radius = square(get<0>(coords));
+    for (size_t i = 1; i < Dim; ++i) {
+      inv_radius += square(coords.get(i));
+    }
+    inv_radius = 1.0 / sqrt(inv_radius);
+    if (type_ ==
+        detail::ConstraintPreservingSphericalRadiationType::Sommerfeld) {
+      get(*dt_pi_correction) = -get(dt_pi);
+      for (size_t i = 0; i < Dim; ++i) {
+        get(*dt_pi_correction) += normal_covector.get(i) * dt_phi.get(i);
+      }
+    } else if (type_ == detail::ConstraintPreservingSphericalRadiationType::
+                            FirstOrderBaylissTurkel) {
+      get(*dt_pi_correction) = -get(dt_pi) + inv_radius * get(dt_psi);
+      for (size_t i = 0; i < Dim; ++i) {
+        get(*dt_pi_correction) += normal_covector.get(i) * dt_phi.get(i);
+      }
+    } else {
+      // second-order Bayliss-Turkel
+      get(*dt_pi_correction) =
+          -get(dt_pi) -
+          2.0 * inv_radius * (2.0 * get(pi) - inv_radius * get(psi));
+      for (size_t i = 0; i < Dim; ++i) {
+        // Note: here we are handling dt drPsi as n^idtPhi_i and dr dtPsi
+        // as -n^id_iPi. The only thing we are assuming is that d_r is
+        // time-independent. This is why we have n^i(dt Phi_i - d_i Pi) instead
+        // of 2n^i dt Phi_i.
+        get(*dt_pi_correction) +=
+            normal_covector.get(i) *
+            (dt_phi.get(i) - d_pi.get(i) + 4.0 * inv_radius * phi.get(i));
+        for (size_t j = 0; j < Dim; ++j) {
+          get(*dt_pi_correction) +=
+              normal_covector.get(i) * normal_covector.get(j) * d_phi.get(i, j);
+        }
+      }
+    }
+  }
+  if (face_mesh_velocity.has_value()) {
+    // Compute dt psi correction
+    get(*dt_psi_correction) =
+        get<0>(normal_covector) * (get<0>(d_psi) - get<0>(phi));
+    for (size_t i = 1; i < Dim; ++i) {
+      get(*dt_psi_correction) +=
+          normal_covector.get(i) * (d_psi.get(i) - phi.get(i));
+    }
+    const auto negative_lambda0 =
```

Comment:
Is lambda_0 = lambda_psi here? If so, perhaps add to docs? In any case, I would find a pointer in the docs useful to how the char speeds relate to the mesh velocity.